lib/realtime/vehicle.ex

```diff
@@ -63,19 +87,136 @@ defmodule Realtime.Vehicle do
-      stop_id: json["vehicle"]["stop_id"]
+      id: id,
+      label: label,
+      timestamp: timestamp,
+      direction_id: direction_id,
+      route_id: route_id,
+      trip_id: trip_id,
+      current_status: current_status,
+      stop_id: stop_id,
+      status: %{
+        stop_granularity: %{
+          current_status: current_status,
+          stop_id: stop_id
+        },
+        timepoint_granularity: %{
+          current_status: timepoint_status(current_status, percent_of_the_way_to_next_timepoint),
+          timepoint_id: next_timepoint.id,
+          percent_of_the_way: percent_of_the_way_to_next_timepoint
+        }
+      }
+    }
+  end
+
+  @spec percent_of_the_way_to_next_timepoint([Stop.id()], [Timepoint.t()], Stop.id()) :: {
+          non_neg_integer(),
+          Timepoint.t()
+        }
+
+  def percent_of_the_way_to_next_timepoint(stop_ids, _timepoints, _stop_id)
+      when is_list(stop_ids) and length(stop_ids) == 0,
+      do: {0, @empty_timepoint}
+
+  def percent_of_the_way_to_next_timepoint(stop_ids, timepoints, stop_id) do
+    {past_stop_ids, future_stop_ids} = split_stops(stop_ids, stop_id)
+
+    next_timepoint_stop_id =
+      Enum.find([stop_id | future_stop_ids], &is_a_timepoint?(&1, timepoints))
+
+    next_timepoint =
+      Enum.find(timepoints, &(&1.stop_id == next_timepoint_stop_id)) || @empty_timepoint
+
+    count_to_previous_timepoint = count_to_timepoint(past_stop_ids, timepoints)
+    count_to_next_timepoint = count_to_timepoint(future_stop_ids, timepoints)
+
+    {
+      percent_of_the_way(count_to_previous_timepoint, count_to_next_timepoint),
+      next_timepoint
     }
   end
 
+  @spec split_stops([Stop.id()], Stop.id()) :: {[Stop.id()], [Stop.id()]}
+  defp split_stops(stop_ids, stop_id) do
+    stop_index = Enum.find_index(stop_ids, &(&1 == stop_id))
+    split_stops_on_index(stop_ids, stop_index)
+  end
+
+  @spec split_stops_on_index([Stop.id()], non_neg_integer() | nil) :: {[Stop.id()], [Stop.id()]}
+  defp split_stops_on_index(_stop_ids, nil), do: {[], []}
+
+  defp split_stops_on_index(stop_ids, index) do
+    {past, [_current | future]} = Enum.split(stop_ids, index)
+
+    {Enum.reverse(past), future}
+  end
+
+  @spec count_to_timepoint([Stop.id()], [Timepoint.t()]) :: non_neg_integer()
+  defp count_to_timepoint(stop_ids, timepoints) do
+    count = Enum.find_index(stop_ids, &is_a_timepoint?(&1, timepoints))
+
+    if is_number(count), do: count + 1, else: 0
+  end
+
+  @spec is_a_timepoint?(Stop.id(), [Timepoint.t()]) :: boolean
+  defp is_a_timepoint?(stop_id, timepoints), do: Enum.any?(timepoints, &(&1.stop_id == stop_id))
+
+  @spec percent_of_the_way(non_neg_integer(), non_neg_integer()) :: non_neg_integer()
+  defp percent_of_the_way(past_count, future_count) do
+    if past_count + future_count == 0 do
+      0
+    else
+      Kernel.trunc(
+        past_count / (past_count + future_count) *
```

Comment:
We could get the twice the resolution by counting `stopped_at` and `incoming_at` separately for each stop. This would also remove the need for a `timepoint_status` field to tell if a bus is exactly at a timepoint, since a bus would only be `100%` if it's `boarding_at` the stop associated with that timepoint.